File: apps/ios/RunningDays/Core/Utilities/Formatters.swift
import Foundation

// MARK: - Formatters
// Utility functions matching the web app's formatting (src/lib/utils.ts)

enum Formatters {
    // MARK: - Distance

    /// Format distance in meters to km string
    /// - Parameter meters: Distance in meters
    /// - Returns: Formatted string like "5.42" or "12.3"
    static func formatDistance(_ meters: Double) -> String {
        let km = meters / 1000
        if km >= 100 {
            return String(format: "%.0f", km)
        } else if km >= 10 {
            return String(format: "%.1f", km)
        }
        return String(format: "%.2f", km)
    }

    // MARK: - Duration

    /// Format duration in seconds to time string
    /// - Parameter seconds: Duration in seconds
    /// - Returns: Formatted string like "45:32" or "1:32:15"
    static func formatDuration(_ seconds: Int) -> String {
        let hours = seconds / 3600
        let minutes = (seconds % 3600) / 60
        let secs = seconds % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, secs)
        }
        return String(format: "%d:%02d", minutes, secs)
    }

    /// Format duration in seconds to hours string
    /// - Parameter seconds: Duration in seconds
    /// - Returns: Formatted string like "4.5" hours
    static func formatDurationHours(_ seconds: Int) -> String {
        let hours = Double(seconds) / 3600
        if hours >= 10 {
            return String(format: "%.0f", hours)
        }
        return String(format: "%.1f", hours)
    }

    // MARK: - Pace

    /// Format pace from seconds per km to min:sec string
    /// - Parameter secondsPerKm: Pace in seconds per kilometer
    /// - Returns: Formatted string like "5:42"
    static func formatPace(_ secondsPerKm: Double) -> String {
        guard secondsPerKm > 0 && secondsPerKm.isFinite else { return "--:--" }

        let minutes = Int(secondsPerKm) / 60
        let seconds = Int(secondsPerKm) % 60
        return String(format: "%d:%02d", minutes, seconds)
    }

    /// Calculate pace from distance and duration
    /// - Parameters:
    ///   - distanceMeters: Distance in meters
    ///   - durationSeconds: Duration in seconds
    /// - Returns: Pace in seconds per kilometer
    static func calculatePace(distanceMeters: Double, durationSeconds: Int) -> Double? {
        guard distanceMeters > 0 else { return nil }
        let distanceKm = distanceMeters / 1000
        return Double(durationSeconds) / distanceKm
    }

    // MARK: - Ordinal

    /// Format number with ordinal suffix
    /// - Parameter number: The number to format
    /// - Returns: Formatted string like "1st", "2nd", "3rd", "4th"
    static func ordinal(_ number: Int) -> String {
        let suffix: String
        let onesDigit = number % 10
        let tensDigit = (number / 10) % 10

        if tensDigit == 1 {
            suffix = "th"
        } else {
            switch onesDigit {
            case 1: suffix = "st"
            case 2: suffix = "nd"
            case 3: suffix = "rd"
            default: suffix = "th"
            }
        }

        return "\(number)\(suffix)"
    }

    // MARK: - Date

    /// Format date to relative string
    /// - Parameter date: The date to format
    /// - Returns: Formatted string like "Today", "Yesterday", "Mon, Dec 27"
    static func formatRelativeDate(_ date: Date) -> String {
        let calendar = Calendar.current

        if calendar.isDateInToday(date) {
            return "Today"
        } else if calendar.isDateInYesterday(date) {
            return "Yesterday"
        } else {
            let formatter = DateFormatter()
            formatter.dateFormat = "EEE, MMM d"
            return formatter.string(from: date)
        }
    }

    /// Format date to short string
    /// - Parameter date: The date to format
    /// - Returns: Formatted string like "Dec 27"
    static func formatShortDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }

    // MARK: - Percentage

    /// Format percentage with optional decimal places
    /// - Parameters:
    ///   - value: The percentage value (0-100)
    ///   - decimals: Number of decimal places
    /// - Returns: Formatted string like "82%" or "82.5%"
    static func formatPercent(_ value: Double, decimals: Int = 0) -> String {
        String(format: "%.\(decimals)f%%", value)
    }
}

File: packages/utils/src/formatters.ts
/**
 * Format distance in kilometers
 * @param meters - Distance in meters
 * @returns Formatted string like "5.23" or "12.3"
 */
export function formatDistance(meters: number): string {
  const km = meters / 1000;
  return km < 10 ? km.toFixed(2) : km.toFixed(1);
}

/**
 * Format duration from seconds to HH:MM:SS or MM:SS
 * @param seconds - Duration in seconds
 * @returns Formatted string like "1:23:45" or "23:45"
 */
export function formatDuration(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
  }
  return `${minutes}:${secs.toString().padStart(2, '0')}`;
}

/**
 * Format pace (min/km) from speed (m/s)
 * @param metersPerSecond - Speed in meters per second
 * @returns Formatted string like "5:30" (minutes:seconds per km)
 */
export function formatPace(metersPerSecond: number): string {
  if (metersPerSecond <= 0) return '--:--';
  const secondsPerKm = 1000 / metersPerSecond;
  const minutes = Math.floor(secondsPerKm / 60);
  const seconds = Math.floor(secondsPerKm % 60);
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
}

/**
 * Format pace from seconds per kilometer
 * @param secondsPerKm - Pace in seconds per kilometer
 * @returns Formatted string like "5:30" (minutes:seconds per km)
 */
export function formatPaceFromSeconds(secondsPerKm: number | null): string {
  if (!secondsPerKm || secondsPerKm <= 0) return '--:--';
  const minutes = Math.floor(secondsPerKm / 60);
  const seconds = Math.floor(secondsPerKm % 60);
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
}

/**
 * Calculate pace from distance and duration
 * @param distanceMeters - Distance in meters
 * @param durationSeconds - Duration in seconds
 * @returns Speed in meters per second
 */
export function calculatePace(distanceMeters: number, durationSeconds: number): number {
  if (distanceMeters <= 0 || durationSeconds <= 0) return 0;
  return distanceMeters / durationSeconds;
}

/**
 * Get ordinal suffix for a number (1st, 2nd, 3rd, etc.)
 * @param n - The number
 * @returns Number with ordinal suffix
 */
export function ordinal(n: number): string {
  const s = ['th', 'st', 'nd', 'rd'];
  const v = n % 100;
  return n + (s[(v - 20) % 10] || s[v] || s[0]);
}

/**
 * Format a percentage with specified decimal places
 * @param value - Decimal value (0-1) or percentage (0-100)
 * @param decimals - Number of decimal places
 * @param isDecimal - Whether the input is a decimal (true) or percentage (false)
 */
export function formatPercent(value: number, decimals = 0, isDecimal = true): string {
  const percentage = isDecimal ? value * 100 : value;
  return `${percentage.toFixed(decimals)}%`;
}
